### ISSUE TITLE
Fortinet access / support

### ISSUE THREAD

**Issue by noblomov** [ytti/oxidized#640]
Hello,

I have a working oxidized settings, backuping configurations for F5 and Cisco boxes.
I don't manage to backup my fortinet boxes though.

The output I have frmo the logs is :
_W, [2016-12-13T14:01:24.134470 #28459]  WARN -- : 10.162.114.9 raised Timeout::Error with msg "execution expired"
W, [2016-12-13T14:01:44.135712 #28459]  WARN -- : 10.162.114.9 raised Net::OpenTimeout (rescued Timeout::Error) with msg "timed out while opening a connection to the host"_

My config file for the fortinets are as follow 
IP_address:fortios:admin:admin

Any help wold be appreciated as I hardly find any info on fortinet & the issue I am getting.

Thanks.

**Comment by ytti** [comment]
You might want to include input debugging:
```
input:
  debug: true
```

**Comment by noblomov** [comment]
Log file attached.

[log_Fortinet.txt](https://github.com/ytti/oxidized/files/649032/log_Fortinet.txt)

**Comment by ytti** [comment]
So there is 'more' page prompt as last line? 

It seems to end '--More-- ', we expect end-of-line ```$```, try to remove that, see what happens.

Remove dollar from here:
https://github.com/ytti/oxidized/blob/master/lib/oxidized/model/fortios.rb#L7

**Comment by noblomov** [comment]
same problem is reported...I'll try and dig the code ; if you have any other hint that would be helpful.

**Comment by AnotherIdea** [comment]
I'm currently working on this problem @ work

Could we deactivate Paging with post_login and reactivate it with pre_logout ?
Tested with fortios 5.0 and 5.4
`config system console
set output standard
end`

and
`config system console
set output more
end`

May probably not work with vdom still...

**Comment by ytti** [comment]
Sure disabling paging would be useful.  Just bit odd why the pager isn't working. Is it always stoping at the same spot?

**Comment by AnotherIdea** [comment]
Just tried it. Seems to be a timeout issue... Getting different trace as I change the timeout value
Fortigate in the same LAN as my test machine.

Any idea why it's slow with oxidized and fast with direct ssh ?

**Comment by ytti** [comment]
Can you arrange access to sanitised device from 91.198.120.1 and send authentication details to saku@ytti.fi. Would be easier to troubleshoot with access to device.

**Comment by noblomov** [comment]
Hi. disabling paging with the commands proposed by AnotherIdea worked. I haven't tested yet with vdom though.

Here is what I have in the fortios.rb conf :

```
  cfg :telnet, :ssh do
    post_login 'config system console'
    post_login 'set output standard'
    post_login 'end'

    pre_logout 'config system console'
    pre_logout 'set output more'
    pre_logout 'end'

    pre_logout "exit\n"

```

**Comment by ytti** [comment]
Oh hmm, we probably don't want to much about with config, that does not seem kosher. I'd try to get the pager prompt handling fixed.

**Comment by noblomov** [comment]
that would be indeed cleaner. Let us know when you are able to have this fixed.
Thanks !

**Comment by ytti** [comment]
Alas I don't know what is the problem with it. Either someone else fixes it or provides me with access from 91.198.120.1 to affected device.

**Comment by rbeumer** [comment]
I had the same issue, first it worked fine and suddenly there were logs about time outs. Nothing was changed and we fixed it by disabling paging as posted by noblomov.

**Comment by ecm75** [comment]
I have the same problem with Fortinet firewalls. I'm able to work around this by raising the timeout value in config to 240 seconds but that is not optimal because now it takes this long for unreachable boxes to fail...
I'd expect the timeout to count when there is no answer from the box but here it seems it is more like a execution timer timeout which fires after X seconds although the box is responding ans pushing it's config.

On a side note: Is it possible to override the timeout value per node?

**Comment by ytti** [comment]
Does this timeout happen with ssh or telnet, or both? We should only timeout if any single command takes over timeout. There shouldn't be timeout for the whole session, if there is, there is a bug and it clearly should be larger than command timeout.

I think optionally passing timeout per node is useful, and we definitely should add that. But first I want to understand the problem better.

**Comment by ecm75** [comment]
I only tested ssh, also I have seen this bahaviour only with Fortinet Nodes so far.

**Comment by ytti** [comment]
Can you test telnet too. Also it would be great if we'd have some understanding on the timing, I understood that you suspect that the device isn't pausing ever for long time, so you feel like there is some timer triggering even though we're constantly getting output from device? If this is true, we have a bug.

You could do:
```
input:
  debug: true
```

then ```tail -f <file>``` during fetch to get some in-sight into timing. 

Or you could copy the model into ```~/config/oxidized/model/``` directory, and modify it so that you get timestamp on the fetched config. Like:

```
cmd "something" do |output|
  output = [Time.now, output].join "\n"
  #normal output mangle
end
```

To see if there is significant delay between some commands in the output.

**Comment by ecm75** [comment]
Not sure if I can test telnet on the device as well, I will have a look...

But I will test the debug code and get back to you.

**Comment by ecm75** [comment]
Ok, I tested both ssh and telnet:
1. telnet works as expected (and is quite fast)
Although the debug log when connecting with telnet doesn't show the expect regex
2. SSH doesn't work and I think I saw the problem in debug log:
it's expecting the prompt instead of the '--More--' string when doiung 'show' command 
```D, [2017-01-24T13:17:34.597846 #15503] DEBUG -- : lib/oxidized/worker.rb: Added X.X.X.X to the job queue
D, [2017-01-24T13:17:34.598372 #15503] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-01-24T13:17:34.598278 #15503] DEBUG -- : lib/oxidized/job.rb: Starting fetching process for X.X.X.X at 2017-01-24 12:17:34 UTC
D, [2017-01-24T13:17:34.599422 #15503] DEBUG -- : lib/oxidized/input/ssh.rb: Connecting to X.X.X.X
D, [2017-01-24T13:17:34.839416 #15503] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/^([-\w\.\~]+(\s[\(\w\-\.\)]+)?\~?\s?[#>$]\s?)$/] at X.X.X.X
D, [2017-01-24T13:17:35.598642 #15503] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-01-24T13:17:35.914542 #15503] DEBUG -- : lib/oxidized/input/cli.rb: Running post_login commands at X.X.X.X
D, [2017-01-24T13:17:35.914796 #15503] DEBUG -- : lib/oxidized/model/model.rb Collecting commands' outputs
D, [2017-01-24T13:17:35.915004 #15503] DEBUG -- : lib/oxidized/model/model.rb Executing get system status
D, [2017-01-24T13:17:35.915199 #15503] DEBUG -- : lib/oxidized/input/ssh.rb get system status @ X.X.X.X with expect: /^([-\w\.\~]+(\s[\(\w\-\.\)]+)?\~?\s?[#>$]\s?)$/
D, [2017-01-24T13:17:35.915609 #15503] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/^([-\w\.\~]+(\s[\(\w\-\.\)]+)?\~?\s?[#>$]\s?)$/] at X.X.X.X
D, [2017-01-24T13:17:36.598929 #15503] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-01-24T13:17:37.518819 #15503] DEBUG -- : lib/oxidized/model/model.rb Executing config global
D, [2017-01-24T13:17:37.519016 #15503] DEBUG -- : lib/oxidized/input/ssh.rb config global @ X.X.X.X with expect: /^([-\w\.\~]+(\s[\(\w\-\.\)]+)?\~?\s?[#>$]\s?)$/
D, [2017-01-24T13:17:37.519277 #15503] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/^([-\w\.\~]+(\s[\(\w\-\.\)]+)?\~?\s?[#>$]\s?)$/] at X.X.X.X
D, [2017-01-24T13:17:37.599248 #15503] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-01-24T13:17:38.345939 #15503] DEBUG -- : lib/oxidized/model/model.rb Executing get hardware status
D, [2017-01-24T13:17:38.346152 #15503] DEBUG -- : lib/oxidized/input/ssh.rb get hardware status @ X.X.X.X with expect: /^([-\w\.\~]+(\s[\(\w\-\.\)]+)?\~?\s?[#>$]\s?)$/
D, [2017-01-24T13:17:38.346495 #15503] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/^([-\w\.\~]+(\s[\(\w\-\.\)]+)?\~?\s?[#>$]\s?)$/] at X.X.X.X
D, [2017-01-24T13:17:38.599543 #15503] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-01-24T13:17:39.523435 #15503] DEBUG -- : lib/oxidized/model/model.rb Executing end
D, [2017-01-24T13:17:39.523623 #15503] DEBUG -- : lib/oxidized/input/ssh.rb end @ X.X.X.X with expect: /^([-\w\.\~]+(\s[\(\w\-\.\)]+)?\~?\s?[#>$]\s?)$/
D, [2017-01-24T13:17:39.523918 #15503] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/^([-\w\.\~]+(\s[\(\w\-\.\)]+)?\~?\s?[#>$]\s?)$/] at X.X.X.X
D, [2017-01-24T13:17:39.599865 #15503] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-01-24T13:17:40.073681 #15503] DEBUG -- : lib/oxidized/model/model.rb Executing show
D, [2017-01-24T13:17:40.073876 #15503] DEBUG -- : lib/oxidized/input/ssh.rb show @ X.X.X.X with expect: /^([-\w\.\~]+(\s[\(\w\-\.\)]+)?\~?\s?[#>$]\s?)$/
D, [2017-01-24T13:17:40.074158 #15503] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/^([-\w\.\~]+(\s[\(\w\-\.\)]+)?\~?\s?[#>$]\s?)$/] at X.X.X.X
```

here is the same box with telnet:
```D, [2017-01-24T13:27:25.584568 #15630] DEBUG -- : lib/oxidized/worker.rb: Added X.X.X.X to the job queue
D, [2017-01-24T13:27:25.584620 #15630] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-01-24T13:27:25.584968 #15630] DEBUG -- : lib/oxidized/job.rb: Starting fetching process for X.X.X.X at 2017-01-24 12:27:25 UTC
D, [2017-01-24T13:27:25.933120 #15630] DEBUG -- : lib/oxidized/input/cli.rb: Running post_login commands at X.X.X.X
D, [2017-01-24T13:27:25.933288 #15630] DEBUG -- : lib/oxidized/model/model.rb Collecting commands' outputs
D, [2017-01-24T13:27:25.933387 #15630] DEBUG -- : lib/oxidized/model/model.rb Executing get system status
D, [2017-01-24T13:27:25.933484 #15630] DEBUG -- : Telnet: get system status @X.X.X.X
D, [2017-01-24T13:27:26.159121 #15630] DEBUG -- : lib/oxidized/model/model.rb Executing config global
D, [2017-01-24T13:27:26.159298 #15630] DEBUG -- : Telnet: config global @X.X.X.X
D, [2017-01-24T13:27:26.199079 #15630] DEBUG -- : lib/oxidized/model/model.rb Executing get hardware status
D, [2017-01-24T13:27:26.199286 #15630] DEBUG -- : Telnet: get hardware status @X.X.X.X
D, [2017-01-24T13:27:26.264525 #15630] DEBUG -- : lib/oxidized/model/model.rb Executing end
D, [2017-01-24T13:27:26.264758 #15630] DEBUG -- : Telnet: end @X.X.X.X
D, [2017-01-24T13:27:26.308002 #15630] DEBUG -- : lib/oxidized/model/model.rb Executing show
D, [2017-01-24T13:27:26.308225 #15630] DEBUG -- : Telnet: show @X.X.X.X
D, [2017-01-24T13:27:26.584888 #15630] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-01-24T13:27:27.585314 #15630] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-01-24T13:27:28.585837 #15630] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-01-24T13:27:29.586160 #15630] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-01-24T13:27:30.586482 #15630] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-01-24T13:27:31.586824 #15630] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-01-24T13:27:32.589701 #15630] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-01-24T13:27:33.590112 #15630] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-01-24T13:27:34.590497 #15630] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-01-24T13:27:35.590874 #15630] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-01-24T13:27:36.592534 #15630] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-01-24T13:27:37.601133 #15630] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-01-24T13:27:38.601410 #15630] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-01-24T13:27:39.604523 #15630] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-01-24T13:27:40.604877 #15630] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2017-01-24T13:27:41.511586 #15630] DEBUG -- : lib/oxidized/input/cli.rb Running pre_logout commands at X.X.X.X
D, [2017-01-24T13:27:41.511832 #15630] DEBUG -- : Telnet: exit
 @X.X.X.X
D, [2017-01-24T13:27:41.567401 #15630] DEBUG -- : lib/oxidized/node.rb: Oxidized::Telnet ran for X.X.X.X successfully
```

**Comment by ecm75** [comment]
Here's another thing I noticed while debugging: 
The resulting config (in plaintext and git) contains some control characters every few lines, probably where the --More-- propt was shown. In less it a reverse ^M.
I haven't noticed this with other models yet and I think the config might not be copy&paste usable with these characters...

**Comment by ecm75** [comment]
I have to revert some statements from the post with the logs:

the regex is fine and also the --More-- expect works!

But oxidized runs into timeout while processing the config from the fortigate and processes the --More-- pager messages. Using ssh this takes a lot of time (in debug you can see it "printing the lines to the file...)
config dump on my node using telnet takes ~15-16secs while using ssh it takes ~7 Minutes (with input debug, without it's faster but still >3mins.
So the problem here is the processing of the main config dump command (show) and pager processing takes longer than the timeout.
Is it supposed to work like this? I'd expect the timeout counter to be reset eg. on pager action (because something is still going on and the node answeres).

**Comment by ytti** [comment]
If you change these https://github.com/ytti/oxidized/blob/master/lib/oxidized/input/ssh.rb#L141-L142 to say 0.01, do you see improvement?

**Comment by ecm75** [comment]
It's a bit faster but doesn't really makes a difference. The fortigate seems to be just slower dumping it's config over ssh than telnet. I even set 0.001 which doesn't speed up the process any further.

Is it possible to reset the timeout timer when the pager expect hits? I think this would circumvent the timeout problem for nodes just being slow and/or having large config files.

**Comment by ytti** [comment]
Do you experience the slowness over CLI SSH too? I.e. without oxidized telnet is fast, ssh is slow, that is, we cannot fix the slowness issue in oxidized. 

For expect prompt handling, we actually do not 'refresh' timer, it is seen as single command, and prompt must be found before timer expires, so each specific command must complete within timer. We could potentially change this, but is it safe, what if there is forever loop in the device, that prompt is never presented?

But I like the idea of per-device timeout, I think it makes whole lot of sense.

**Comment by ecm75** [comment]
Ok, I see the problem now and your point with the loop.
I see a slight difference in speed between ssh and telnet without oxidized but it's not that big of a difference..

I think the node timeout would be the best solution here and solve the main problem of raising the global timeout to values like 300s for all nodes. It's quite acceptable to have just a few nodes with high timeout values.

**Comment by ytti** [comment]
If it's really slow on oxidized, but not on SSH CLI, then something is wrong with oxidized. Would it be possible to arrange me access to sanitised device exhibiting from this problem? I'd come from 91.198.120.1. 
Oxidized shouldn't take significantly longer than CLI.

**Comment by ecm75** [comment]
I will check if we have a spare device which we could setup for debugging.

**Comment by ecm75** [comment]
We have a spare device which I will configure for you to test with, I'm hoping to get to it tomorrow.
I will let you know when it is ready (and if it shows the same bahaviour...)

**Comment by zeki893** [comment]
to get my fortigates working all I had to do was set timeout to 3600. Config is pretty long with multiple vdoms. I tail the log file and I can see it is pulling and takes a long time to complete.

**Comment by ecm75** [comment]
Hey ytti,

I'm sorry it took so long but I finally managed to setup a fortigate for you to test with.
Where can I send you the access credentials?

**Comment by ytti** [comment]
saku@ytti.fi please

**Comment by arturs0** [comment]
I'm also experiencing the same issue. 

As additional input when I use SSH to backup a fortigate in the same network segment as the oxidized server the issue does not appear. When trying to backup configuration over WAN then this issue appears.

**Comment by udhos** [comment]
Is this issue being worked on? In my environment, Oxidized is very slow to fetch FortiOS configuration over SSH. I guess that disabling the pager is a practical work-around.

**Comment by ytti** [comment]
I think I may have received the auth, I'm lagging on my end.

**Comment by ecm75** [comment]
No problem, is the timeout override per device still on your agenda or should we open a separate issue for this request for feature?

**Comment by ytti** [comment]
Yeah it's not lot of work to do the per node timeout, just seems to be too much anyhow.

**Comment by mvangoor** [comment]
Any update on this issue?

**Comment by udhos** [comment]
@mvangoor The work-around https://github.com/ytti/oxidized/issues/640#issuecomment-267044391 tackled the issue for me.

**Comment by mvangoor** [comment]
I currently increased the overall timeout waiting for a final fix

**Comment by denngie** [comment]
I would just like to add that with paging it took 18 minutes to get the config from our Fortigate with a conf of 66k lines with paging, without ~30 sec. And with paging we got blank lines where ---more--- was replaced.

**Comment by admiralspark** [comment]
@denngie which version of FortiOS? Just checked my different installs and 5.6 is the only one with spaces so far. Can you verify you're running a 5.6 box?

**Comment by denngie** [comment]
@admiralspark We are running 5.2.11. Config:
```
config system global
    set admintimeout 30
    set fgd-alert-subscription advisory latest-threat
    set gui-dlp enable
    set gui-endpoint-control disable
    set gui-explicit-proxy enable
    set gui-wireless-controller disable
    set hostname &quot;fgate&quot;
    set refresh 5
    set service-expire-notification disable

         
    set timezone 26
    set vdom-admin enable
end
config system accprofile
    edit &quot;prof_admin&quot;
        set mntgrp read-write
        set admingrp read-write
        set updategrp read-write
        set authgrp read-write
        set sysgrp read-write
        set netgrp read-write
        set loggrp read-write
        set routegrp read-write
        set fwgrp read-write
        set vpngrp read-write
        set utmgrp read-write
        set wanoptgrp read-write
        set endpoint-control-grp read-write
        set wifi read-write
    next
    edit &quot;read_only&quot;
        set mntgrp read

         
        set admingrp read
        set updategrp read
        set authgrp read
```
We don't have any devices running 5.6 unfortunately

**Comment by admiralspark** [comment]
Interesting, I'm on 5.2.11 and don't have that, but my 5.6.2 definitely does. Oh well, thought I had an idea...

**Comment by jsynack** [comment]
The latest updates to the FortiOS model use the `show full-configuration | grep .` method to bypass the pager within FortiOS.

This should make things run smoother without the need to toggle any config bits during the oxidized polling of the device.

Can you test / validate with the latest FortiOS model from the git repo?

**Comment by henriots** [comment]
jsynack solution worked for me

**Comment by jerebernard** [comment]
Hello 
I have a Fortiweb device and the "show full-configuration | grep ."  create in my case an error.
This is the result : 
# # COMMAND: get system status
# International Version: FortiWeb-600D 5.83,build1388,170728
# Serial-Number: xxxxxxxxxx
# Bios version: 00010001
# Log hard disk: Available
# Hostname: xxxxxxxxxxx
# Operation Mode: Reverse Proxy
# FIPS-CC mode: disabled
# Current HA mode: standalone
# 
# # COMMAND: get hardware status
# Parsing error at 'hardware'. err=1
# Command fail. CLI parsing error.
# 
 
# COMMAND: show full-configuration | grep .
Parsing error at '|'. err=1
Command fail. CLI parsing error.
